incubator/hnc/pkg/reconcilers/hierarchy_config.go

```diff
@@ -145,6 +156,70 @@ func (r *HierarchyConfigReconciler) reconcile(ctx context.Context, log logr.Logg
 	return r.updateObjects(ctx, log, nm)
 }
 
+// syncDeleting returns true if the namespace or the HC instance is being deleted.
+func (r *HierarchyConfigReconciler) syncDeleting(ctx context.Context, log logr.Logger, inst *api.HierarchyConfiguration, nsInst *corev1.Namespace) (bool, error) {
```

Comment:
General comment - this function's really complicated, it has a lot of different cases and it's hard for me to understand how they all fit together. I think it's fine for now since when we move the namespace creation/deletion to the HNSR, this will get a lot simpler, but it's probably worth a comment.

In the future, I think you can simplify this by checking all the `DeletionTimestamp.IsZero()` conditions _outside_ the "sync" function, and then call different "sync" functions for each one. Eg:

```
switch {
case nsInst.CreationTimestamp.IsZero():
  return r.syncMissingNamespace()
case nsInst.DeletionTimestamp.IsZero():
  return r.syncDeletingNamespace()
case inst.CreationTimestamp.IsZero():
  return r.syncMissingHC()
case inst.DeletionTimestamp.IsZero():
  return r.syncDeletingHC()
default:
  return r.syncHC()
}
```

You can keep this in mind when making the next few PRs.